systemd/systemd.go

```diff
@@ -692,12 +705,56 @@ func IsTimeout(err error) bool {
 	return isTimeout
 }
 
+func (l Log) parseLogRawMessageString(key string, sliceHandler func([]string) (string, error)) (string, error) {
+	valObject, ok := l[key]
+	if !ok || valObject == nil {
+		// NOTE: journalctl says that sometimes if a json string would be too
+		// large null is returned, so we may miss a message here
+		return "", fmt.Errorf("key %q missing from message", key)
+	}
+
+	// first try normal string
+	s := ""
+	err := json.Unmarshal(*valObject, &s)
+	if err == nil {
+		return s, nil
+	}
+
+	// try runes next, this is the case if journald thinks the output is not
+	// safe ascii
+	r := []rune{}
+	err = json.Unmarshal(*valObject, &r)
```

Comment:
https://www.freedesktop.org/wiki/Software/systemd/json/ states:

> A field that contains non-printable or non-UTF8 is serialized as a number array instead. This is necessary to handle binary data in a safe way without losing data, since JSON cannot embed binary data natively. Each byte of the binary field will be mapped to its numeric value in the range 0..255.

If we have invalid utf8 here, we should hit the error in line 743 eventually.

Reply:
I refactored to parse into a []byte, and then let Go figure out what to do about invalid UTF-8. I think it's fine to use Go's escaping here for invalid UTF-8 since this is used for `snap logs` which is essentially user-facing and so providing something that is readable is better than providing numbers etc. Users/programs concerned with accuracy or dealing with known binary data should probably be talking with journald directly instead of going through snapd like this.